### PR TITLE
Bytecode soundness patch on last byte

### PR DESCRIPTION
**Pull request by rrzhang139**
### Description

Fixes bug soundness in bytecode circuit, where an attacker can insert as many bytes after the last byte (row where `q_last`=1). 

### Issue Link

https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1332

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- Bug fix 1: Add a constraint that forces the next state to be "Header" when the current state is "Byte" and `index == length - 1`
- Bug fix 2: set `q_enable=1` for the row where `q_last=1`

### Rationale

This would patch the possibility of an attacker inserting false bytes that extend beyond the correct bytecode. 

### How Has This Been Tested?

Created a test called `bytecode_soundness_bug_1` that overwrites the last rows with`tag=Byte`. 
`cargo test --features test --package zkevm-circuits --lib -- bytecode_circuit::circuit::tests::bytecode_soundness_bug_1 --exact --nocapture`
Also ran other unit tests to check correctness was maintained.

### CODE REVIEW

**Comment by rrzhang139** [comment]
@ed255 pushed some changes to fix bug 1 and bug 2. unit tests don't pass because I still have your overwrite test which should fail now.

File: circuit-benchmarks/src/bench_params.rs
Comment (ed255) [inline circuit-benchmarks/src/bench_params.rs:1]: This new file seems unrelated to this PR

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ed255) [inline zkevm-circuits/src/bytecode_circuit/circuit.rs:519]:
```suggestion
                // Overwrite the witness assignment by using the values in the `overwrite` parameter.  This is used to explicitly set intermediate witness values for negative tests.
```
I think extending this comment will add more clarity on what's going on here

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ed255) [inline zkevm-circuits/src/bytecode_circuit/circuit.rs:1219]:
```suggestion
    #[test]
    #[should_panic]
    fn bytecode_soundness_bug_1() {
```

See https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic

This allows us to keep this negative test and still pass the CI

**Review by ed255 — COMMENTED** [review]
LGTM overall!  Thanks for submitting the fix to this soundness bug.
My main comment is about tagging the bug test as `should_panic` so that we can keep it and pass the CI.  See my other two small comments as well.

I'll approve the PR once the CI passes

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (leolara) [inline zkevm-circuits/src/bytecode_circuit/circuit.rs:537]: If we are overriding, shouldn't we also override the value_rlc or leave it alone alone as it was before?

Perhaps we could implement a system that allows to choose which columns we override?

**Comment by leolara** [comment]
Perhaps we take this opportunity to alsof fix: https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1347

**Review by leolara — CHANGES_REQUESTED** [review]

File: circuit-benchmarks/src/bench_params.rs
Comment (leolara) [reply circuit-benchmarks/src/bench_params.rs:1]: I agree this change should not be included, please

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ed255) [reply zkevm-circuits/src/bytecode_circuit/circuit.rs:368]: I think I don't understand this proposal.
Would this gate change following your proposal?  Maybe you can make a diff to show how it would look?

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ed255) [reply zkevm-circuits/src/bytecode_circuit/circuit.rs:537]: > If we are overriding, shouldn't we also override the value_rlc or leave it alone alone as it was before?

I wrote this overwrite code originally.  The reason why `value_rlc` is not included in the overwrite values is because it can't be precomputed as it depends on the challange, which is only known at phase 2.  So the lines below compute a correct value_rlc during assignment.

> Perhaps we could implement a system that allows to choose which columns we override?

This is the approach taken in the state circuit: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/0e76d2b25e62c22f75b26a64d9d7ea78bb3fede0/zkevm-circuits/src/state_circuit.rs#L500
Nevertheless I'm not sure if it would be useful here. (I think it would make the negative tests more complex)

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (leolara) [reply zkevm-circuits/src/bytecode_circuit/circuit.rs:368]: I don't know how to add the diff.

So this gate would not exists and we would add `cb.require_zero(and::expr(vec![is_next_byte(meta), index_length_diff_is_zero.clone().is_zero_expression])` to the gate annotated with "Byte Row", in main line starting in 177.

So if you are in a Byte row: next row to be byte AND index_length_diff == 0 must be false.

I think that would follow more the rest of the code structure.

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (leolara) [reply zkevm-circuits/src/bytecode_circuit/circuit.rs:537]: Ok

**Comment by leolara** [comment]
I guess we need to update the spec?

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ed255) [reply zkevm-circuits/src/bytecode_circuit/circuit.rs:368]: Thanks for the diff!
So the original constraint was formed like: If we're in this tag and this condition, then next tag must be this
The new one is formed like: if we're in this tag, then this next tag can't happen with this condition.

They are equivalent, but I think the original one is a bit more clear?  I would leave @rrzhang139 to choose the way forward.

**Comment by rrzhang139** [comment]
@ed255 Should I perhaps remove the override code and the bytecode_soundness_bug_1() test? Or is there a better way to test overriding without changing the function signature? Seems unnecessary to have it just for one test.

**Comment by ed255** [comment]
> @ed255 Should I perhaps remove the override code and the bytecode_soundness_bug_1() test? Or is there a better way to test overriding without changing the function signature? Seems unnecessary to have it just for one test.

I don't follow.  Are you referring to the `assign` and `assign_internal` function signatures?  I think it's not too bad to leave the override code and the bug test, specially because the `overwrite` parameter is not included in the `BytecodeCircuit` constructor.  Let me know if you're referring so something else.

**Comment by leolara** [comment]
We will use override for other tests in the future. But I insist that keeping the "bug 1" or "bug 2" in the code will not bring clarity as nobody will know what these number mean in a few months.

**Review by ed255 — APPROVED** [review]
LGTM!

**Review by leolara — APPROVED** [review]

**Comment by rrzhang139** [comment]
@ed255 @leolara can I get this merged now?

**Comment by leolara** [comment]
@rrzhang139 yes, you can merge with two approvals 👍

**Comment by rrzhang139** [comment]
@leolara I can't merge because I don't have write access on this repo

**Comment by leolara** [comment]
@rrzhang139 sorry I never know who can

**Pull request merged**
